### ISSUE TITLE
Wazuh packages redesign tier 1 - Stage 2 - Migrate and adapt Wazuh subsystem repositories

### ISSUE THREAD

**Issue by MarcelKemp** [wazuh/wazuh#21755]
| Related issue |
| :--: |
| https://github.com/wazuh/internal-devel-requests/issues/712 |

# Description

Stage 2 of the Wazuh packages redesign consists of migrating the code hosted in the [wazuh-packages repository](https://github.com/wazuh/wazuh-packages), adapting it, and adding CI

For this, a series of tasks have been designed to be carried out by the different teams involved.

| Task |
| :-- |
|Migrate Wazuh manager |
|Migrate Wazuh agent |
|Migrate WPK |
|Migrate Filebeat files/module |

Each of these tasks must meet the following objectives:

- Research and set ETA for the parent issue (notify @rauldpm)
- Issue: Migrate code into target repository - Manual package generation
- Issue: Package creation through CI
  - Packages that require a `sign` during the build phase must implement this functionality (for example Windows and macOS)
- Issue: Add secrets to GH (S3 and CloudFront access - QA team)
- Issue: Add CI testing
- Issue: Upload packages to S3 through CI. This must be done after the testing phase

## Functional and non-functional requirements
> [!note]
> See https://github.com/wazuh/internal-devel-requests/issues/187#issue-1871209243

- As subsystem owners, each team is responsible for the package testing. This means that the package installs and passes some basic validations
- As package users, we want to easily identify the package's nature. To do this, each package must be identifiable and unique.
- As subsystem owners, teams must follow this naming convention for development package names: `<SubsystemName>_<VersionNumber>-<Revision>_{<OSName>-<OSVersion>_}<Architecture>_<GitReference>.<PackageType>`.
> OS name and version are optional.
> Inspired by [Debian's package convention](https://www.debian.org/doc/manuals/debian-faq/pkg-basics.en.html).

  - As subsystem owners, we want to use tools efficiently. This implies that whenever the requested package already exists in the storage, the generation request will be skipped.
  - Authorized users and automation tools will be able to generate a package by providing:
    - the subsystem
    - package type (deb, rpm, ...)
    - system architecture
    - optional: operating system name and version
    - a valid Git commit

  - Successfully built packages will be automatically uploaded to the storage.
  - The package-building process must log the result of the task and provide the final package URL.
--------------------------------------------
  - The subsystem's version and revision will be extracted from its metadata. The Dockerfiles must extract that data and make it accessible within the container.

  - The building of the packages must be done through the use of Docker containers.
  - The building of the packages must be automated using GitHub Actions.
  - Once a package is successfully built, it must be automatically uploaded to the Wazuh package storage (S3 bucket or other).
  - If GitHub Actions doesn't provide an appropriate runner for a particular package, external runners will be used (AWS, Site Ox, MacStadium, etc.). These runners deployment and connection must be as similar as possible as the current deployment and connection processes used in Jenkins.
  - The packages' generation GH Action execution must be standardized between subsystems.
  - The Filebeat module is owned by the core team and they must create a GitHub action to generate the tar.gz file.

## Pre-requisites
- https://github.com/wazuh/wazuh-jenkins/issues/6180 (partially)

> [!NOTE]
> Packages must be uploaded to `packages-dev.internal.wazuh.com` bucket. The final path is pending to be validated

# Tasks

- [Resources related](https://github.com/wazuh/wazuh-packages) 
- [x] https://github.com/wazuh/wazuh/issues/22320 (debs, rpms)
- [x] https://github.com/wazuh/wazuh/issues/22080 (arch, debs, macos, rpms,  windows)
  - [x] https://github.com/wazuh/wazuh/issues/22868
- [x] https://github.com/wazuh/wazuh/issues/22839
- [x] https://github.com/wazuh/wazuh/issues/22321
- [x] https://github.com/wazuh/wazuh/issues/22367
- [x] https://github.com/wazuh/wazuh-agent-packages/issues/17
- [ ] https://github.com/wazuh/wazuh-documentation/issues/7184

## Sub-tasks for each task

- [ ] Research and set ETA for the parent issue (notify @rauldpm)
- [ ] Migrate code into target repository - Manual package generation
- [ ] Package creation through CI
  - Packages that require a sign during the build phase must implement this functionality (for example Windows and macOS)
- [ ] Add CI testing
- [ ] Upload packages to S3 through CI. This must be done after the testing phase

# Approved by

- DRI name: @rauldpm

**Comment by rauldpm** [comment]
- Estimated work duration per issue: 
  - Migrate and adapt the Wazuh manager subsystem (debs, rpms) (~2 weeks)
  - https://github.com/wazuh/wazuh/issues/22080#issuecomment-1964645576 (~2 weeks)
  - Migrate and adapt WPK (~1 week)
  - Migrate and adapt the Filebeat templates (~1 week)
- Estimated work duration:
  - Sequential (worst case - 1 assigned): 6 weeks
  - Parallel (best case - 4 assigned): 2 weeks
- Provisional ETA based on worst case (starting to be worked at `2024/03/04`): `2024/04/12`

**Comment by rauldpm** [comment]
The repository has been configured with CI secrets, please, refer to the main comment of the https://github.com/wazuh/wazuh-jenkins/issues/6290 issue to check the secrets names, the CI used has been configured to put and get objects from the development warehouse

**Comment by rauldpm** [comment]
Issue to migrate the Windows sign VM to AWS EC2: https://github.com/wazuh/internal-devel-requests/issues/1004

**Comment by rauldpm** [comment]
Related to:

- https://github.com/wazuh/wazuh/issues/22320
- https://github.com/wazuh/wazuh/issues/22080

After discussing it with @wazuh/devel-devops, it was decided to use the Allocation module since using the AWS CLI it would be necessary to make several calls to obtain connection information and so on. In addition, the module will also be used for macOS (intel/arm) and Windows (and in the future other packages like PowerPC)

Thus, workflows that build ARM/Windows packages must:
1. Clone the wazuh-qa repository using the same workflow branch/tag
2. Access the module directory
3. Create a YAML inventory with the desired deployment information
4. Run the module Python script with inventory
5. Process the YAML provided by the script (get the connection information)
6. Use this information to connect to the machines
7. Perform the corresponding actions to generate the package

Once the whole process is finished, the remote machine must be destroyed using the same script and make sure that it is (this last one is very important)

Remember that the VPN connection is needed to access the desired resources in MacStadium/EC2

On the part of QA, we will provide the CI user with EC2 permissions for Windows once the module is finished

For now, until the Allocation module is finished, these tasks cannot be continued for ARM/Windows (only sign tasks for Windows, the package build is not affected by this)

The use of the Allocation module will be provided through the Allocation module documentation (available when merged into a numbered branch)
